File: ios/Classes/model/map/overlay/overlay/NPolygonOverlay.swift
import NMapsMap

internal struct NPolygonOverlay: AddableOverlay {
    typealias OverlayType = NMFPolygonOverlay
    var overlayPayload: Dictionary<String, Any?> = [:]

    let info: NOverlayInfo
    let coords: Array<NMGLatLng>
    let color: UIColor
    let holes: Array<Array<NMGLatLng>>
    let outlineColor: UIColor
    let outlineWidth: Double
    let outlinePatternPx: Array<NSNumber>

    func createMapOverlay() -> OverlayType {
        let polygon = NMGPolygon(
            ring: NMGLineString(points: coords),
            interiorRings: holes.map({ NMGLineString(points: $0) })
        )
        let overlay = NMFPolygonOverlay(polygon as! NMGPolygon<AnyObject>)!
        return applyAtRawOverlay(overlay)
    }
    
    func applyAtRawOverlay(_ overlay: NMFPolygonOverlay) -> NMFPolygonOverlay {
        overlay.fillColor = color
        overlay.outlineColor = outlineColor
        overlay.outlineWidth = UInt(round(outlineWidth))
        overlay.outlinePattern = outlinePatternPx
        return overlay
    }

    static func fromMessageable(_ v: Any) -> NPolygonOverlay {
        let d = asDict(v)
        return NPolygonOverlay(
                info: NOverlayInfo.fromMessageable(d[infoName]!),
                coords: asArr(d[coordsName]!, elementCaster: asLatLng),
                color: asUIColor(d[colorName]!),
                holes: asArr(d[holesName]!, elementCaster: { asArr($0, elementCaster: asLatLng) }),
                outlineColor: asUIColor(d[outlineColorName]!),
                outlineWidth: asDouble(d[outlineWidthName]!),
                outlinePatternPx: asArr(d[outlinePatternName]!) {
                    NSNumber(value: DisplayUtil.dpToPx(dp: asDouble($0)))
                }
        )
    }

    /*
        --- Messaging Name Define ---
    */
    private static let infoName = "info"
    static let coordsName = "coords"
    static let colorName = "color"
    static let holesName = "holes"
    static let outlineColorName = "outlineColor"
    static let outlineWidthName = "outlineWidth"
    static let outlinePatternName = "outlinePattern"
    static let boundsName = "bounds"
}

File: android/src/main/kotlin/dev/note11/flutter_naver_map/flutter_naver_map/model/map/overlay/overlay/NPolygonOverlay.kt
package dev.note11.flutter_naver_map.flutter_naver_map.model.map.overlay.overlay

import androidx.annotation.ColorInt
import com.naver.maps.geometry.LatLng
import com.naver.maps.map.overlay.PolygonOverlay
import dev.note11.flutter_naver_map.flutter_naver_map.model.map.overlay.overlay.AddableOverlay
import dev.note11.flutter_naver_map.flutter_naver_map.converter.DefaultTypeConverter.asDouble
import dev.note11.flutter_naver_map.flutter_naver_map.converter.DefaultTypeConverter.asInt
import dev.note11.flutter_naver_map.flutter_naver_map.converter.DefaultTypeConverter.asList
import dev.note11.flutter_naver_map.flutter_naver_map.converter.DefaultTypeConverter.asMap
import dev.note11.flutter_naver_map.flutter_naver_map.converter.MapTypeConverter.asLatLng
import dev.note11.flutter_naver_map.flutter_naver_map.model.map.info.NOverlayInfo
import dev.note11.flutter_naver_map.flutter_naver_map.util.DisplayUtil

internal data class NPolygonOverlay(
    override val info: NOverlayInfo,
    val coords: List<LatLng>,
    @ColorInt val color: Int,
    val holes: List<List<LatLng>>,
    @ColorInt val outlineColor: Int,
    val outlineWidthDp: Double,
    val outlinePatternDp: List<Double>,
) : AddableOverlay<PolygonOverlay>() {



    override fun createMapOverlay(): PolygonOverlay = applyAtRawOverlay(PolygonOverlay())

    override fun applyAtRawOverlay(overlay: PolygonOverlay) = overlay.also { g ->
        g.coords = coords
        g.color = color
        g.holes = holes
        g.outlineColor = outlineColor
        g.outlineWidth = DisplayUtil.dpToPx(outlineWidthDp)
        g.setOutlinePattern(*outlinePatternDp.map { DisplayUtil.dpToPx(it) }.toIntArray())
    }

    companion object {
        fun fromMessageable(rawMap: Any): NPolygonOverlay = rawMap.asMap().let {
            NPolygonOverlay(
                info = NOverlayInfo.fromMessageable(it[infoName]!!),
                coords = it[coordsName]!!.asList { latLng -> latLng.asLatLng() },
                color = it[colorName]!!.asInt(),
                holes = it[holesName]!!.asList { hole ->
                    hole.asList { latLng -> latLng.asLatLng() }
                },
                outlineColor = it[outlineColorName]!!.asInt(),
                outlineWidthDp = it[outlineWidthName]!!.asDouble(),
                outlinePatternDp = it[outlinePatternName]!!.asList { p -> p.asDouble() },
            )
        }

        /*
            --- Messaging Name Define ---
        */

        private const val infoName = "info"
        const val coordsName = "coords"
        const val colorName = "color"
        const val holesName = "holes"
        const val outlineColorName = "outlineColor"
        const val outlineWidthName = "outlineWidth"
        const val outlinePatternName = "outlinePattern"
        const val boundsName = "bounds"
    }
}